gomap/gomap.gobra

```diff
@@ -0,0 +1,121 @@
+/*
+  This file is part of gobra-libs which is released under the MIT license.
+  See LICENSE or go to https://github.com/viperproject/gobra-libs/blob/main/LICENSE
+  for full license details.
+*/
+
+// This package defines Go maps and their associated operations in terms of
+// mathematical maps (dictionaries).
+package gomap
+
+import "dicts"
+
+// Converts a Go map to a mathematical map (dictionary).
+ghost
+requires acc(m, _)
+ensures domain(result) == domain(m)
+ensures forall k int :: {result[k]} k in domain(m) ==> result[k] == m[k]
+decreases
+pure func ToDict(m map[int]int) (result dict[int]int)
+
+// Returns a new map.
+ensures acc(result)
+ensures ToDict(result) == dicts.Empty()
+func Make() (result map[int]int) {
+	return make(map[int]int)
+}
+
+// Updates the key to the value.
+requires acc(m)
+ensures acc(m)
+ensures ToDict(m) == ToDict(old(m))[k = v]
+func Update(m map[int]int, k, v int) {
+	m[k] = v
+}
+
+// Returns the value stored under the key.
+requires acc(m, _)
+ensures k in domain(m) ==> result == ToDict(m)[k]
+ensures !(k in domain(m)) ==> result == 0
+decreases
+pure func Get(m map[int]int, k int) (result int) {
+	return m[k]
+}
+
+// Returns the value stored under the key and whether it exists.
+requires acc(m, 1/2)
+ensures acc(m, 1/2)
+ensures k in domain(m) ==> value == ToDict(m)[k] && ok
+ensures !(k in domain(m)) ==> value == 0 && !ok
+// Cannot be pure, since pure functions must have exactly on result argument.
+func GetExists(m map[int]int, k int) (value int, ok bool) {
+	return m[k]
```

Comment:
This function does not seem necessary to me. A pure function to see if the key is in the domain of the map would be a lot more practical

Reply:
I don't think it is possible to have a non-ghost pure function that does this. The best I could come up with is `Exists(m, k)`.

Reply:
You can implement it as a ghost pure function, which makes it usable in specs:

```go
ghost
requires acc(m, 1/2)
ensures k in domain(m) ==> ok
ensures !(k in domain(m)) ==> !ok
decreases
pure func Contains(m map[int]int, k int) (ok bool) {
	return k in domain(m)
}
```